### PR TITLE
feat: add initial telegram logic

### PR DESCRIPTION
**Pull request by wxiaoyun**
## Description

This PR introduces the initial Telegram integration and WebSocket setup for the backend. It includes the implementation of several Telegram bot commands with placeholder responses for AI interactions and sets up a simple, single-purpose WebSocket for real-time communication with the admin dashboard. Additionally, boilerplate code for the `dataaccess` package is included.

### WebSocket Integration
- **Endpoint**: The admin dashboard connects to the backend WebSocket at `<base_url>/api/ws`.
- **Authentication**: This endpoint requires a JWT token for authentication.
- **Functionality**: The WebSocket is currently unidirectional, with the backend emitting messages to connected clients. 
- **Message format**: Messages are transmitted in json byte format
     ```go
     type WebSocketMessage struct {
         Type string `json:"type"`
         Data any    `json:"data"`
     }
     ```
  - For authentication requests, type field will be `"auth"`
      ```json
      {
         "type": "auth",
         "data": {
            "chat_id": 123,
            "credentials": "super_secret_credentials"
         }
      }
      ```
  - For real time message updates, type field will be `"message_received"`. Note timestamp is in RFC3339 format. Dashboard can use the `chat_id` field to determine whether this message is relevant
      ```json
      {
         "type": "message_received",
         "data": {
            "chat_id": 123,
            "request_query_id": 7
            "telegram_message_id": 345,
            "by": "bot",
            "message_body": "whats 9 plus 10",
            "timestamp": "2006-01-02T15:04:05Z07:00"
         }
      }
      ```

#### Telegram Messages
- Prerequisites
  - Started the chat with `/start`
  - [optional] Authenticated, if making a request

- Flow
  1. All messages received by backend will be catergorized by AI.
  2. Creates the corresponding request query.
  3. Broadcasts the message after, (message have a non-null requestquery field)
  4. Fetch ai response
  5. Broadcasts that response

#### Telegram Bot Commands

1. **`/help`**
   - **Description**: Lists all available commands along with a brief description of each.
   - **Response**: A message containing the list of commands and their descriptions.

2. **`/auth`**
   - **Description**: User attempts to get authenticated by sending their credentials. The backend relays these credentials to all connected dashboards via WebSocket in JSON format.
   - **Example Command**: `/auth super_secret_credentials`
   - **Response**: A message indicating the credentials have been sent for authentication.

3. **`/query`**
   - **Description**: Submits a query regarding the hotel. This command does not require prior authentication.
   - **Response**: Placeholder content for hotel query responses.

4. **`/request`**
   - **Description**: Submits a logistical request. This command requires prior authentication using `/auth`.
   - **Response**: Placeholder content for logistical request responses.

- **NOTE** 
  - `/query and /request` are escape hatches for both guest and developer. Guest can directly make query/request in case the AI fails to correctly categorize. Developer can directly test the workflow of query and request with these commands.
  - `/help, /auth` messages will NOT be broadcasted via web-socket since all these commands has potential to be run without starting a query/request (required by db schema).

### Telegram Authentication
- To authenticate a specific chat, the dashboard should make a `POST` request to `<base_url>/api/tg/auth` with the following payload:
    ```go
    type Params struct {
        ChatID     uint   `json:"chat_id" validate:"required"`
        RoomNumber string `json:"room_number" validate:"required"`
        LastName   string `json:"last_name" validate:"required"`
    }
    ```
- Once received, the backend will send a success message to that chat.

### Boilerplate Code

- **`dataaccess` Package**: Added initial boilerplate code to support future database interactions and data management.

### CODE REVIEW

File: internal/telegram/chatbot.go [hunk 1]
```diff
@@ -1,34 +1,74 @@
...
+		bot.Debug = true
+	}
 
 	updateConfig := tgbotapi.NewUpdate(0)
-
 	updateConfig.Timeout = 30
 
 	updates := bot.GetUpdatesChan(updateConfig)
+	handleUpdates(bot, hub, updates)
+}
```
Comment (seelengxd) [inline internal/telegram/chatbot.go:37]:
```suggestion
	updateConfig := tgbotapi.NewUpdate(0)
	updateConfig.Timeout = 30

	setCommands(bot)

	updates := bot.GetUpdatesChan(updateConfig)
	handleUpdates(bot, hub, updates)
}

func setCommands(bot *tgbotapi.BotAPI) {
	commandsConfig := tgbotapi.NewSetMyCommandsWithScopeAndLanguage(
		tgbotapi.NewBotCommandScopeAllPrivateChats(),
		"en",
		tgbotapi.BotCommand{Command: tgmessagehandler.HelpCmdWord, Description: tgmessagehandler.HelpCmdDesc},
		tgbotapi.BotCommand{Command: tgmessagehandler.AuthCmdWord, Description: tgmessagehandler.AuthCmdDesc},
		tgbotapi.BotCommand{Command: tgmessagehandler.QueryCmdWord, Description: tgmessagehandler.QueryCmdDesc},
		tgbotapi.BotCommand{Command: tgmessagehandler.RequestCmdWord, Description: tgmessagehandler.RequestCmdDesc},
	)

	msg, err := bot.Request(commandsConfig)
	log.Info(msg)
	if err != nil {
		log.Fatalf("error setting commands: %s", err.Error())
	}
}
```

image

(small addition - add command config)

for future reference: i just figured out that all configs we can send are a `Chattable` and we can use bot.Request to set them :^D

did not include /start

**Review by seelengxd — COMMENTED** [review]
great work as usual! sorry for the late review, will continue tonight/tmr morning
i have a few questions so far

File: internal/telegram/chatbot.go (hunk 1 above)
Comment (ForAeons) [reply internal/telegram/chatbot.go:37]: That a great addition

File: internal/model/message.go [hunk 2]
```diff
@@ -17,14 +17,14 @@ const (
...
-	TelegramMessageId int64
+	TelegramMessageID int64
 	By                By
 	MessageBody       string
 	Timestamp         time.Time
 	HotelStaffId      *uint
 	HotelStaff        *User
 	RequestQueryId    uint
-	RequestQuery      RequestQuery
+	RequestQuery      *RequestQuery
```
[the remark this answers is not in the document]
Comment (ForAeons) [reply internal/model/message.go:27]: Just a small optimization and also making the model structs more consistent.

- Use message struct use less memory. If `RequestQuery` is not a pointer, everytime Message struct is instantiated, Go allocates memory for a RequestQuery as well.
- In terms of sematics, having Message composed of RequestQuery is kinda weird also. See [Golang composition](https://www.geeksforgeeks.org/composition-in-golang/)

File: internal/handler/tgmessagehandler/start.go
[the remark this answers is not in the document]
Comment (ForAeons) [reply internal/handler/tgmessagehandler/start.go:1]: Sure! Will update this

File: internal/model/message.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (seelengxd) [reply internal/model/message.go:27]: i see! had the misguided idea that it meant optional, oops

**Review by seelengxd — APPROVED** [review]
lgtm, thanks!!

File: internal/handler/tgmessagehandler/util.go [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+// TODO: Implement AI response
+func getAICategorisation(_ *gorm.DB, _ string) (model.Type, error) {
+	decision := rand.Intn(3)
+	switch decision {
+	case 0:
+		return model.TypeQuery, nil
+	case 1:
+		return model.TypeRequest, nil
+	default:
+		return model.TypeUnknown, nil
```
Comment (seelengxd) [inline internal/handler/tgmessagehandler/util.go:40]: note: from my understanding, i think the ai does not return this response

i added this as i thought a zero value may be convenient for implementation

**Pull request merged**